# Python for Analytics

Python is a general purpose programming language. It can be used for anything from building a webpage to running the "brain" for autonomous vehicles. This notebook will focus on the specific parts of Python we'll need for data analysis and predictive modeling. Some of this you saw in the initial `IntroToPython` notebook, but this more pointed notebook will hopefully help you troubleshoot some common issues.

## Packages

We'll start with the packages/libraries we'll be utilizing. There are several libraries we'll use during almost every python session. 

+ `pandas` allows us to work with data frames. A data frame is an array of data where columns represent variables and rows represent observations or records. You should think of data frames as single-sheet Excel spreadsheets. These are sometimes called *flat-files* and common formats for this raw data is a `*.csv` or `*.txt` file.
+ `numpy` gives us access to advanced numerical functionality. We'll use it mostly for random number generation.
+ `matplotlib` and `seaborn` are both plotting libraries -- `matplotlib` is great for simple plots, but sometimes `seaborn` is convenient if we want to do something a bit more complex.
+ `sklearn` is the most popular machine learning library on the planet. We'll never import the full `sklearn` library, but we will import parts of it. In particular, `sklearn` gives us the specific functionality necessary to build and assess various classes of model.

### Importing Packages

When we initialize a python session, we get a *lean* version of the software -- it initializes with only the most basic functionality. If we wish to do anything "specialized", we need to load additional functionality. We do this via `import` statements. You might envision a library, much like our own Wolak Library and Learning Commons, where each room corresponds to a package and contains items on shelves that make up that package. We can import individual items, entire shelves, or the whole room!

We'll run some of our common imports below:

In [ ]:
import pandas as pd #import the entire pandas room -- aliased as pd
import numpy as np #import the entire numpy room -- aliased as np
import matplotlib.pyplot as plt #import the entire pyplot shelf in the matplotlib room -- aliased as plt 
import seaborn as sns #import the entire seaborn room -- aliased as sns

### Aliasing

You'll notice when we imported functionality (entire rooms or shelves), we used an alias (`as xxyy`). This is because when we want to use functionality from a library, we'll need to reference where python can find it. That is, if we call `plt.plot()` python will look for the `plot()` item on the `pyplot` shelf in the `matplotlib` room. This is much more convenient than typing `matplotlib.pyplot.plot()`, although they are functionally equivalent.

## Data

Since we're interested in analytics, we need data to learn from. The data we use will come in multiple forms:

+ Data frames (data organized into rows and columns, like an Excel spreadsheet)
+ Columns (an individual column of a data frame)
+ Series (lists of values, not organized as a column)

Knowing the *shape* and *type* of the data we are working with will be really helpful in troubleshooting some of the errors we will encounter.

### Reading Data

Let's read in some data below.

In [ ]:
measels = pd.read_csv("https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-25/measles.csv")
measels.head()

The code cell above did two things:

1. Read in the `measels` data from the provided URL
  + You can think of this as asking Python to take out a carboard box, write `measels` on the front of it, and then dump the results of `pd.read_csv()` into the box so that we can access it later.

2. Print out the `head()` of the `measels` data
  + You can think of this as asking python to go get that box with `measels` written on the front of it, and then show you the first five rows (which is what the `head()` function asks for).

Notice that the `measels` data is arranged as a data frame. It seems as though each row represents a particular school district during a given academic year. The columns are measured variables (related to the mmr vaccine) on those districts and years.

### Subsetting Data

There are three ways we can subset data frames. We can subset by column, by row, or by both column and row. You saw a bit on this when you worked through the initial introduction to python notebook. If we want to subset by row and column simultaneously, we'll need the `.loc()` or `.iloc()` methods, but typically we can get away without them.

#### Subsetting Columns

Let's start by just isolating the state column.

In [3]:
measels.head()

index    state     year  ...  xper        lat         lng
0      1  Arizona  2018-19  ...   NaN  31.347819 -110.938031
1      2  Arizona  2018-19  ...   NaN  32.221922 -110.896103
2      3  Arizona  2018-19  ...   NaN  32.130493 -111.117005
3      4  Arizona  2018-19  ...   NaN  33.485447 -112.130633
4      5  Arizona  2018-19  ...  2.33  33.495620 -112.224722

[5 rows x 16 columns]

In [4]:
measels["state"]

0          Arizona
1          Arizona
2          Arizona
3          Arizona
4          Arizona
           ...    
66108    Wisconsin
66109    Wisconsin
66110    Wisconsin
66111    Wisconsin
66112    Wisconsin
Name: state, Length: 66113, dtype: object

In [5]:
type(measels["state"])

pandas.core.series.Series

Notice that this prints out as a column, actually Python is thinking of `measels["state"]` as a pandas Series object. You can verify this by wrapping the code above in the `type()` function. Try it -- edit the code cell above and then run it with `Shift+Enter` to see the result.

We can easily select multiple columns as well. We just need to pass the *list* of column names we want to call rather than just a single column name. Passing a *list* is important.

In [6]:
measels[["state", "year", "enroll", "mmr"]].head()

state     year  enroll    mmr
0  Arizona  2018-19    51.0  100.0
1  Arizona  2018-19    22.0  100.0
2  Arizona  2018-19    85.0  100.0
3  Arizona  2018-19    60.0  100.0
4  Arizona  2018-19    43.0  100.0

Notice that we used double-brackets in the code above. The inner brackets define the *list* of column names we are requesting. We again use the `.head()` method to print out the first five rows of the object we've retreived.

**Aside (*the "dot" notation*):** You've noticed by this point that we have used commands of the form `object.action()`, for example `measels.head()`. This convenient notation allows us to apply the action on the right of the dot to the object on the left of the dot. We can use this convention to chain commands together -- just remember that the object to the left of the dot will be transformed by the action on the right of the dot.

Now, let's say we wanted to know whether all 50 states are contained in the `measels` data frame. We'll go through a series of attempts before getting a clean answer.

In [12]:
#Isolate the state column
measels["state"]

0          Arizona
1          Arizona
2          Arizona
3          Arizona
4          Arizona
           ...    
66108    Wisconsin
66109    Wisconsin
66110    Wisconsin
66111    Wisconsin
66112    Wisconsin
Name: state, Length: 66113, dtype: object

Hmmm, 66113 rows of data to comb through (remember Python starts counting from 0) -- no thank you! Let's look at the unique values appearing in this column. The `unique()` function takes a column of data and only returns the unique values in that column.

In [18]:
measels["state"].unique()

array(['Arizona', 'Arkansas', 'California', 'Colorado', 'Connecticut',
       'Florida', 'Idaho', 'Illinois', 'Iowa', 'Maine', 'Massachusetts',
       'Michigan', 'Minnesota', 'Missouri', 'Montana', 'New Jersey',
       'New York', 'North Carolina', 'North Dakota', 'Oklahoma', 'Ohio',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Dakota',
       'Tennessee', 'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'Wisconsin'], dtype=object)

Notice that the format of the data being printed out was different than when we asked to just isolate the `state` column. The type of object resulting from these two requests are different, and this can cause frustrating challenges if we are not paying close attention. We can get the same object type by converting to a `Series` using `pandas` (see below). This takes some getting used to, but aside from spelling and capitalization issues, data type errors are the most common errors we'll experience in QSO370/570.

In [20]:
pd.Series(measels["state"].unique())

0            Arizona
1           Arkansas
2         California
3           Colorado
4        Connecticut
5            Florida
6              Idaho
7           Illinois
8               Iowa
9              Maine
10     Massachusetts
11          Michigan
12         Minnesota
13          Missouri
14           Montana
15        New Jersey
16          New York
17    North Carolina
18      North Dakota
19          Oklahoma
20              Ohio
21            Oregon
22      Pennsylvania
23      Rhode Island
24      South Dakota
25         Tennessee
26             Texas
27              Utah
28           Vermont
29          Virginia
30        Washington
31         Wisconsin
dtype: object

The `index` of the series does some counting for us. Using the `index` we notice that only 32 states were included (remember python starts counting at 0). If we hadn't converted to a Series, we could still get python to count how many unique states there are in the dataset using the `len()` function. Unfortunately chaining `.len()` does not work (there are reasons why, but they require a deeper understanding of Computer Science than we need here).

In [16]:
len(measels["state"].unique())

32

So we have our answer! Not all of the states are included -- only 32 of them. I wonder if `New Hampshire` was included.

In [21]:
print("New Hampshire" in measels["state"].unique()) #remember, if we want to print multiple outputs, we need the print() function
print("Vermont" in measels["state"].unique())

False
True


New Hampshire isn't included in the dataset but `Vermont` is included. Let's say we only wanted to work with data from that state. We can ask python to take out a new carboard box, write `Vermont` on the front of it, and then dump in only those rows of the `measels` data frame corresponding to `Vermont`. After doing that, we'll show the first eight rows of data from that box -- just making sure that we get what's expected. Notice the use of the "double-equals" sign here (`==`) which denotes a *test for equality* rather than an assignment declaring that two values are equal.

In [ ]:
Vermont = measels[measels["state"] == "Vermont"]
Vermont.head(8)

Looks good to me! You'll notice that the index (those bold numbers on the side) is messed up now -- actually, the index values were inherited from the original `measels` data frame. We want to begin back at row 0, not at 59452 -- inevitably we'll run into challenges with mismatched indices in our course. We can reset the index to start counting at row 0 by using the `.reset_index()` method. Passing the argument `drop = True`, prevents the original index from being appended to the data frame as a new column. Additionally, the `inplace = True` argument saves the change to the data frame rather than making a temporary change and then immediately forgetting it.

In [ ]:
Vermont.reset_index(inplace = True, drop = True)
Vermont.head()

Maybe we didn't just want to focus on Vermont -- let's build out a dataset corresponding to all of the New England States instead.

In [ ]:
NewEngland = measels[(measels["state"].isin(["Maine", "New Hampshire", "Vermont", "Massachussetts", "Rhode Island", "Connecticut"]))]
#Note: My first attempt at creating the NewEngland dataframe was wrong -- I got 
#an ambiguity error and had to Google to fix it -- don't be afraid if you need 
#to Google your coding errors, or post to Slack for help!

print(NewEngland.head())
print(NewEngland["state"].unique()) #still no New Hampshire

Notice that I spelled `Massachusetts` wrong. Because of this error, it doesn't show up as one of the states in the `NewEngland` dataset. Fix my mistake and re-run the code cell above.

We've got a lot of columns here -- some, like `index` are columns that provide no information. Let's say we care about the `state`, `year`, `enroll`, and `mmr` variables. We'll create a smaller data frame by dropping those unnecessary columns below. The default for the `.drop()` method is to drop rows -- since we want to drop columns we change the axis that we are dropping along (`axis = 1`). Again we use the `inplace = True` argument to make the change permanent.

In [ ]:
NewEngland.drop(["index", "name", "type", "city", "district", "overall", "xrel", "xmed", "xper", "lat", "lng"], axis = 1, inplace = True)
NewEngland.head()

Notice that if you run the previous code cell a second time you'll get an error. This is because those columns we wanted to drop are no longer part of the `NewEngland` data frame. Sometimes this can cause confusion, so it can be advantageous to remove the `inplace = True` argument and just store the result in a brand new data frame. That is, you might run: 
 
  `NewEnglandSmall = NewEngland.drop(["index", "name", "type", "city", "district", "overall", "xrel", "xmed", "xper", "lat", "lng"], axis = 1)` 

instead of what we ran previously.

## Computing on Data Frames and Objects

Let's say we wanted to compute the average value for the `mmr` column of the dataset. We can do this with the `.mean()` or `.median()` methods. Remember that, in the presence of outliers the median is a preferred measure of center.

In [ ]:
NewEngland["mmr"].mean()

Try changing the code above to compute the median rather than the mean. What do you notice? Do you think there are outliers in our dataset? Additionally, we should notice that the quantities we are computing are not the average percentage of school-aged children who have gotten the MMR vaccine -- this is the average district-wide percentage -- that is, a district with 100 students is weighted just as heavily in this computation as a district with 15,000 students. We should always take care to understand the quantities we are computing.

Notice that if we try to apply methods designed to summarise numerical data to non-numeric columns, we'll get a type error.

In [ ]:
NewEngland["county"].mean()
#NewEngland["county"].value_counts()

The `.mean()` method cannot be applied to strings. Now, comment out the line asking for the mean and uncomment the line asking for value counts (a frequency table). Re-run the code cell to see and appropriate summary of this column.

Notice that the `.value_counts()` method can be applied to a pandas Series object (a column of a data frame) but not to a numpy array. Recall earlier that we looked at the results of `measels["state"]` and `measels["state"].unique()`; we observed that the printouts were different and saw that the object types were in fact different. Notice in the next code block, running `NewEngland["county"].unique.value_counts()` throws an error. Try running `pd.Series(NewEngland["county"].unique()).value_counts()` instead -- just be super careful with parentheses (and the result is quite boring).

In [ ]:
NewEngland["county"].unique().value_counts()

### Computing Grouped Quantities

What follows is somewhat more advanced, but tends to be quite useful. What if we wanted to compute a quantity (say a median) for multiple groups of data -- say one for each year? We could save out individual datasets for each year and then work with those, but this is really inefficient and inconvenient. The `.groupby()` method becomes quite useful here.

In the code cell below, we select only the five columns we are interested in. We group by `state`, `county`, and `year`, so that when we summarize (via the `.agg()` method) we'll get just a single row for each county per year. Notice that grouping by `state` is necessary since both Vermont and Massachusetts have an Essex County -- if we didn't group by `state`, Essex County, MA and Essex County, VT would fall in the same group. 

In [ ]:
NewEngland[["state", "county", "year", "mmr", "enroll"]].groupby(["state", "county", "year"], as_index = False).agg("median")

Hey -- New Hampshire is listed as a county in Vermont. Looks like we found an issue with our dataset!

## Summary

Okay -- that's enough for now. Here you got a bit more familiar with the types of objects we'll be dealing with most often in QSO370 -- data frames and columns. You saw that python is picky and that it is important to be aware of the *types* of data we are working with. We purposely ran into some issues -- first we encountered simple mis-spellings and next we encountered *type errors*, where a function could not be applied to our object in its current state. We explored the "dot notation" a bit which allows for the object to the left of the dot to be transformed by the operation on the right of the dot. We also saw an example where the dot notation isn't allowed, but we didn't go into why that is -- it's beyond the scope of our course. 

Hopefully you feel a little bit more comfortable with using python to interact with data after moving through this notebook. You shouldn't feel like an "expert" yet, and you won't feel like one even after completing this course. 

## Try on your own

Let's test your command of the material in notebooks 3 and 4 by trying a few short tasks on the `measels` data.

1. Use `.shape` to determine the number of observations and variables contained in the `measels` data frame.
2. We saw that New Hampshire was "incorrectly" listed as a county in the state of Vermont. Create a `NewHampshire` data frame which consists of only the rows in the `measels` data frame where the `county` is `New Hampshire`. How many observations are there? Do you have any guesses about what might be the case here?
3. Create a new data frame called `lowVaxxRate` which contains only the counties (for each year), where the MMR vaccination rate is below 20% (`mmr` is less than 20). You'll also want to restrict that the `mmr` column should not contain the value -1.0 since that seems to indicate no recorded data. How many such districts are there?
4. $\star$Challenge$\star$ Create a data set called `largeDistricts` which consists of only those districts whose enrollment is in the top 1% of enrollments in the dataset. Sort the data frame by the values in the `mmr` column in ascending order. Which of these districts has the lowest overall MMR vaccination rate?